editoast/src/views/authz.rs

```diff
@@ -53,6 +57,50 @@ enum NoSuchUserError {
     NoSuchUser { user_id: i64 },
 }
 
+#[derive(Debug, serde::Serialize, utoipa::ToSchema)]
+#[cfg_attr(test, derive(serde::Deserialize, PartialEq))]
+struct WhoamiResponse {
+    id: i64,
+    name: String,
+    roles: Vec<Role>,
+}
+
+#[utoipa::path(
+    get,
+    path = "",
+    tag = "authz",
+    responses((
+        status = 200,
+        description = "Get the info of the current user",
+        body = inline(WhoamiResponse),
+    ))
+)]
+async fn whoami(Extension(auth): AuthenticationExt) -> Result<Json<WhoamiResponse>> {
+    // This is especially necessary as this endpoint is always queried by the frontend
+    // when loading. Making it return 401 results in a blank page.
+    // We return `Admin` as when no authorization is enabled, we want everyone to have
+    // access to full feature set of OSRD.
+    if let Authentication::SkipAuthorization { name, .. } = auth {
+        // TODO: don't return -1 and a hardcoded name, return a different schema instead, requires frontend changes
+        return Ok(Json(WhoamiResponse {
+            id: -1,
+            name: name.unwrap_or_else(|| "OSRD user".to_string()),
+            roles: Vec::from([Role::Admin]),
+        }));
+    }
+
+    let authorizer = auth.authorizer()?;
+    let user_roles = authorizer
+        .user_roles()
+        .await
+        .map_err(AuthzError::Authorizer)?;
+    Ok(Json(WhoamiResponse {
+        id: authorizer.user_id(),
+        name: authorizer.user_name().to_owned(),
+        roles: user_roles.into_iter().collect(),
+    }))
```

Comment:
We can now simplify this with something like this:

```suggestion
      Ok(Json(WhoamiResponse {
        id: auth.user_id()?.unwrap_or(-1), // This already exists
        name: auth.user_name()?.unwrap_or("OSRD user".into()) // This function must be created like `user_id`.
        roles: auth.user_roles()?, // Same here the return type should be something like `Result<HashSet<Role>, AuthorizationError>`
    }))
```

The idea behind this suggestion is to limit the usage of the `authorizer` as much as possible. What do you think?

Reply:
2cc3bd402

But I'm not a huge fan of inventing data (-1, "OSRD user") just to match a schema. I think that whoami should return a special value when we're skipping the authorization. Otherwise the clients will build on top of these magic values and things can get brittle very fast.

Fine by me for now, we have better things to do, but I think that implementation should change eventually.